torch/csrc/jit/passes/quantization.cpp

```diff
@@ -87,8 +168,80 @@ void InsertObserverNodes(std::shared_ptr<Graph>& graph, Node* observer_node) {
+  std::vector<std::pair<Value*, Node*>> quantInputs;
+  // For storing quantizable values that are output of quantizable nodes
+  // Since same value can go to multiple nodes, we use set so that
+  // we insert quant-dequant node pairs for value only once
+  std::unordered_set<Value*> quantOutputs;
+
+  while (!blocks_to_visit.empty()) {
+    Block* b = blocks_to_visit.top();
+    blocks_to_visit.pop();
+
+    for (Node* n : b->nodes()) {
+      // Schedule the sub blocks
+      for (Block* subblock : n->blocks()) {
+        blocks_to_visit.push(subblock);
+      }
+
+      // We iterate over node inputs to identify which Values
+      // need to be quantized depending on node type
+      for (auto &v : n->inputs()) {
+        if (!v->type()->isSubtypeOf(TensorType::get())) {
+          // Skip quantization for non tensors
+          continue;
+        }
+
+        if (checkIfNodeQuantizable(v->node())) {
+          // Goal of this iteration is to identify the parent node for V
+          // that is quantizable and replace all uses of Value with
+          // quant-dequant output. Usage of set helps adding single
+          // q-dq nodes for all V->users
+          // Example N1 -> (V1 -> (N2), V2 -> (N3))
+          //         N1 is quantizable node. So we insert quant-dequant
+          //         nodes for all outputs of N1 (V1, V2) once
+          quantOutputs.emplace(v);
+        } else if (checkIfNodeQuantizable(n)) {
+          // Goal of this iteration is to identify nodes that are
+          // quantizable but input value originate from non quantizable
+          // node. This requires selectively inserting q-dq nodes for
+          // inputs into node N(V, N pair) because parent node might
+          // also have input into other non quantizable nodes
+          // Example : N1(prim::Param) -> (V1 -> (N4, N5), V2 -> (N6, N7), V3)
+          //           N1 is not quantizable node but N4 and N7 are
+          //           quantizable nodes. So we add the (V1, N4) and
+          //           (V2, N7) as insertion points for quant-dequant nodes
+          quantInputs.emplace_back(std::make_pair(v, n));
+        }
+      }
+    } // End Loop for nodes within block
+
+    // Since we are iterating node inputs only above, we need to iterate
+    // over block outputs values and if they originate from quantizable
+    // node, we push to quantOutputs set
+    auto outputVals = b->outputs();
+    for (auto& v : outputVals) {
+      if (checkIfNodeQuantizable(v->node()) &&
+        v->type()->isSubtypeOf(TensorType::get())) {
+        quantOutputs.emplace(v);
+      }
+    } //end for
+  } // end Block traversal
+
+  // Insert the quant-dequant pair for values output from quantizable nodes
+  for (auto& ele : quantOutputs) {
+    addQuantDeQuantNodes(ele);
+  }
+
+  // Insert the quant-dequant pair for values inputs to quantizable nodes
+  for (auto& ele : quantInputs) {
+    addQuantDeQuantNodesForInput(ele.first, ele.second);
+  }
```

Comment:
Request adding more tests: There is a lot of logic here in terms of checking for quantizable/non-quantizable nodes and in iterating across sub-blocks.